Code/GraphMol/MolDraw2D/MolDraw2D.cpp

```diff
@@ -1130,13 +1205,453 @@ DrawColour MolDraw2D::getColourByAtomicNum(int atomic_num) {
+        line_pts.emplace_back(at1_cds + perp * rad);
+        line_pts.emplace_back(at2_cds + perp * rad);
+        line_pts.emplace_back(at2_cds - perp * rad);
+        line_pts.emplace_back(at1_cds - perp * rad);
+        drawPolygon(line_pts);
+      } else {
+        draw_adjusted_line(at1_cds + perp * rad, at2_cds + perp * rad);
+        draw_adjusted_line(at1_cds - perp * rad, at2_cds - perp * rad);
+      }
+    } else {
+      double col_rad = 2.0 * rad / hb.second.size();
+      if(drawOptions().fillHighlights) {
+        Point2D p1 = at1_cds - perp * rad;
+        Point2D p2 = at2_cds - perp * rad;
+        vector<Point2D> line_pts;
+        for (size_t i = 0; i < hb.second.size(); ++i) {
+          setColour(hb.second[i]);
+          line_pts.clear();
+          line_pts.emplace_back(p1);
+          line_pts.emplace_back(p1 + perp * col_rad);
+          line_pts.emplace_back(p2 + perp * col_rad);
+          line_pts.emplace_back(p2);
+          drawPolygon(line_pts);
+          p1 += perp * col_rad;
+          p2 += perp * col_rad;
+        }
+      } else {
+        int step = 0;
+        for(size_t i = 0; i < hb.second.size(); ++i) {
+          setColour(hb.second[i]);
+          // draw even numbers from the bottom, odd from the top
+          Point2D offset = perp * (rad - step * col_rad);
+          if (!(i % 2)) {
+            draw_adjusted_line(at1_cds - offset, at2_cds - offset);
+          } else {
+            draw_adjusted_line(at1_cds + offset, at2_cds + offset);
+            step++;
+          }
+        }
+      }
+    }
+    setLineWidth(orig_lw);
+  }
+
+}
+
+// ****************************************************************************
+int MolDraw2D::getHighlightBondWidth(int bond_idx,
+                                     const map<int, int> *highlight_linewidth_multipliers) const {
+
+  int bwm = drawOptions().highlightBondWidthMultiplier;
+  // if we're not doing filled highlights, the lines need to be narrower
+  if(!drawOptions().fillHighlights) {
+    bwm /= 2;
+    if(bwm < 1) {
+      bwm = 1;
+    }
+  }
+
+  if(highlight_linewidth_multipliers && !highlight_linewidth_multipliers->empty()) {
+    auto it = highlight_linewidth_multipliers->find(bond_idx);
+    if(it != highlight_linewidth_multipliers->end()) {
+      bwm = it->second;
+    }
+  }
+  int tgt_lw = lineWidth() * bwm;
+
+  return tgt_lw;
+
+}
+
+// ****************************************************************************
+void MolDraw2D::adjustLineEndForHighlight(int at_idx,
+                                          const map<int, double> *highlight_radii,
+                                          Point2D p1, Point2D &p2) const {
+
+  // this code is transliterated from
+  // http://csharphelper.com/blog/2017/08/calculate-where-a-line-segment-and-an-ellipse-intersect-in-c/
+  // which has it in C#
+  double xradius, yradius;
+  Point2D centre;
+  calcLabelEllipse(at_idx, highlight_radii, centre, xradius, yradius);
```

Comment:
the code below will misbehave if either xradius or yradius are 0. Maybe worth testing for that here?